ast/walk.go

```diff
@@ -72,6 +72,122 @@ func Walk(expr Expression, visitor func(Expression) bool) bool {
 	panic(fmt.Sprintf("Unexpected AST expression of type %T: %[1]v", expr))
 }
 
+func ValidateParentAccess(expr Expression) error {
+	_, err := walkAndFindParentUsageInvalid(0, expr)
+	return err
+}
+
+// walkAndFindParentUsageInvalid is like Walk() but it returns an error if an invalid
+// parent usage is found.
+// This function is exclusively used to walk the body of a custom defined function.
+// The usage of a parent operator is *not* allowed in the body of a custom defined function
+// if it reaches out of that function's scope. It's valid within a filter or a projection, because
+// we create NewNestedScope for those.
+func walkAndFindParentUsageInvalid(count int, expr Expression) (int, error) {
+	switch e := expr.(type) {
+	case *Parent:
+		if count <= 0 {
+			return count, fmt.Errorf("parent usage is invalid")
```

Comment:
```suggestion
			return count, fmt.Errorf("parent usage is invalid")
```

Hm. Maybe this should return a `ParseError` so that it's possible to determine where in the query there is an error?

The only weird thing here is that `ParseError` is an interface in the _parser_ package with three (completely) identical implementations. I opened a PR here which moves it into `ast` so that we can reuse it here: https://github.com/sanity-io/go-groq/pull/32